utils/core/src/serde/mod.rs

```diff
@@ -0,0 +1,146 @@
+        let source = flatten_slice_elements(self);
+        T::write_batch_into(source, target);
+    }
+}
+
+impl<T: Serializable, const N: usize> Serializable for &Vec<[T; N]> {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let source = flatten_slice_elements(self);
+        T::write_batch_into(source, target);
+    }
+}
+
+impl<T: Serializable> Serializable for &[T] {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        T::write_batch_into(self, target);
+    }
+}
+
+impl<T: Serializable, const N: usize> Serializable for &[[T; N]] {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let source = flatten_slice_elements(self);
+        T::write_batch_into(source, target);
+    }
+}
+
+// DESERIALIZABLE
+// ================================================================================================
+
+/// Defines how to deserialize `Self` from bytes.
+pub trait Deserializable: Sized {
+    // REQUIRED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Reads a sequence of bytes from the provided `source`, attempts to deserialize these bytes
+    /// into `Self`, and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The `source` does not contain enough bytes to deserialize `Self`.
+    /// * Bytes read from the `source` do not represent a valid value for `Self`.
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError>;
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Attempts to deserialize the provided `bytes` into `Self` and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The `bytes` do not contain enough information to deserialize `Self`.
+    /// * The `bytes` do not represent a valid value for `Self`.
+    ///
+    /// Note: if `bytes` contains more data than needed to deserialize `self`, no error is
+    /// returned.
+    fn read_from_bytes(bytes: &[u8]) -> Result<Self, DeserializationError> {
+        Self::read_from(&mut SliceReader::new(bytes))
+    }
+
+    /// Reads a sequence of bytes from the provided `source`, attempts to deserialize these bytes
+    /// into a vector with the specified number of `Self` elements, and returns the result.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The `source` does not contain enough bytes to deserialize the specified number of
+    ///   elements.
+    /// * Bytes read from the `source` do not represent a valid value for `Self` for any of the
+    ///   elements.
+    ///
+    /// Note: if the error occurs, the reader is not rolled back to the state prior to calling
+    /// this function.
+    fn read_batch_from<R: ByteReader>(
+        source: &mut R,
+        num_elements: usize,
+    ) -> Result<Vec<Self>, DeserializationError> {
```

Comment:
Instead of returning a vec, we might want ot return anything that implements `FromIterator`. This will give us great flexibility and allow us to transparently reconstruct complex structures such as maps, files, databases, etc

```rust
fn read_batch_from<R, T>(
    source: &mut R,
) -> Result<T, DeserializationError> where
R: ByteReader,
T: FromIterator<Self>
```

Reply:
I'm not sure how to provide a generic implementation for this - so, leaving it for the future.